File: src/main/java/com/github/wakingrufus/mastodon/ui/login/OauthController.java
package com.github.wakingrufus.mastodon.ui.login;

import com.github.wakingrufus.mastodon.events.OAuthTokenEvent;
import com.sys1yagi.mastodon4j.MastodonClient;
import com.sys1yagi.mastodon4j.api.entity.auth.AppRegistration;
import javafx.event.Event;
import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.scene.control.TextField;
import javafx.scene.web.WebView;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class OauthController {
    private final MastodonClient client;
    private final AppRegistration appRegistration;
    private final String oauthUrl;
    @FXML
    TextField tokenField;
    @FXML
    Button loginButton;
    @FXML
    WebView webView;

    public OauthController(MastodonClient client, AppRegistration appRegistration, String oAuthUrl) {
        this.oauthUrl = oAuthUrl;
        this.client = client;
        this.appRegistration = appRegistration;
    }

    @FXML
    public void initialize() {
        webView.getEngine().load(oauthUrl);
        loginButton.setOnAction((event) -> {
            //   AccessToken accessToken = clientService.getAccessToken(client, appRegistration.getClientId(), appRegistration.getClientSecret(), tokenField.getText());

            OAuthTokenEvent OAuthTokenEvent = new OAuthTokenEvent(tokenField.getText(),
                    event.getTarget(), appRegistration, tokenField.getText(), client);
            Event.fireEvent(event.getTarget(), OAuthTokenEvent);
        });
    }
}

File: src/main/kotlin/com/github/wakingrufus/mastodon/controllers/OAuthController.kt
package com.github.wakingrufus.mastodon.controllers

import com.github.wakingrufus.mastodon.data.OAuthModel
import javafx.fxml.FXML
import javafx.scene.control.Button
import javafx.scene.control.TextField
import javafx.scene.web.WebView
import mu.KLogging

class OAuthController(private val oAuthModel: OAuthModel,
                      private val oAuthurl: String,
                      private val completeOAuth: (OAuthModel) -> Any)
    : Controller<OAuthModel> {
    companion object : KLogging()

    @FXML
    internal var tokenField: TextField? = null
    @FXML
    internal var loginButton: Button? = null
    @FXML
    internal var webView: WebView? = null

    @FXML
    override fun initialize() {
        webView?.engine?.load(oAuthurl)
        loginButton?.setOnAction { event ->
            completeOAuth.invoke(oAuthModel.copy(token = tokenField!!.text))
        }
    }
}
